lib/library_web/controllers/login_controller.ex

```diff
@@ -0,0 +1,159 @@
+defmodule LibraryWeb.LoginController do
+  use LibraryWeb, :controller
+  alias Library.Users
+  @elixir_auth_github Application.get_env(:library, :elixir_auth_github) || ElixirAuthGithub
+
+  @httpoison Application.get_env(:library, :httpoison) || HTTPoison
+
+  @doc """
+    Just shows the user the plain index.html
+  """
+  def index(conn, _params) do
+    render conn, "index.html"
+  end
+
+
+  @doc """
+    Creates a github URL with the scope for getting the user's email and orgs
+    renders the index.html if there is a failure constructing it (can only
+    happen when elixir_auth_github is not configured correctly/env variables
+    are not set), and redirects to github login if the url is correct
+  """
+  def login(conn, _params) do
+
+    case @elixir_auth_github.login_url_with_scope(["user:email", "read:org"]) do
+      {:err, reason} ->
+        # TODO: set internal server error
+        conn = conn
+              |> put_flash(:error, "Something went wrong on our end, try again later!")
+
+        render conn, "index.html"
+      {:ok, url} ->
+        redirect conn, external: url
+    end
+  end
+
+  @doc """
+    This is the route which will be called after the user has authenticated on
+    github, it takes the code from github, and then ElixirAuthGithub goes and
+    gets the user's details. If there's an error, we send the user to the index
+    page, with an error message, if it's succesful, we proceed to get their
+    email and organisations in separate requests to the github api, and then
+    check if they're a member of dwyl. If all of this happens without a hitch,
+    we insert them into the database (or update their entry, if they're already in there). If there's any errors along the way then we redirect the user to
+    the index page with a message about what went wrong.
+  """
+
+  def callback(conn, %{"code" => code}) do
+    case @elixir_auth_github.github_auth(code) do
+      {:error, error} ->
+        conn
+        |> put_flash(:error, "Problem getting info from github, try again later")
+        |> render("index.html")
+      {:ok, user} ->
+        case add_user(user) do
+            {:ok, user} ->
+              conn = conn
+                    |> put_flash(:info, "added to db succesfully")
+                    |> put_session(:user_id, user.id)
+              redirect(conn, to: page_path(conn, :index))
+            {:error, "not a dwyl member"} ->
+              conn = conn
+                     |> put_flash(:error, "Sorry, the library can only be used by members of dwyl!")
+
+              redirect(conn, to: page_path(conn, :index))
+            error ->
+              conn = conn
+                     |> put_flash(:error, "Problem adding user, sorry, try again later!")
+
+              redirect(conn, to: page_path(conn, :index))
```

Comment:
Same with these ones as the comment above. Remove the `conn=` and just pipe it through to the redirects/renders.

Reply:
The only reason I didn't pipe here was due to using `conn` in the `page_path` as well as in the `redirect` function. Refactoring in a slightly different way, will see what you think when I push up.